### PR TITLE
Implementation of JSON Web Tokens

### PR DESCRIPTION
**Pull request by vtdat**

### CODE REVIEW

File: api/auth.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+	if creds.Password != pass || creds.Username != user {
+		a.respondError(rw, apiError{
+			code: http.StatusUnauthorized,
+			err:  fmt.Errorf("invalid user or password"),
+		})
+		return
+	}
+
+	expTime := time.Now().Add(60 * time.Minute)
```
Comment (ntk148v) [inline api/auth.go]: Allow to config it.

File: api/cloud.go [hunk 2]
```diff
@@ -31,6 +31,9 @@ import (
 )
 
 func (a *API) registerCloud(w http.ResponseWriter, req *http.Request) {
+	if req.Method == "OPTIONS" {
```
Comment (ntk148v) [inline api/cloud.go]: Handling CORS preflight request with a wrapper may be a better option.

File: middleware/middleware.go [hunk 3]
```diff
@@ -92,22 +93,55 @@ func (m *Middleware) Logging(next http.Handler) http.Handler {
...
+			if err != nil {
+				if err == jwt.ErrSignatureInvalid {
+					level.Error(m.logger).Log("msg", "Unauthorized request")
+					http.Error(w, "Login Required!", http.StatusUnauthorized)
+					return
+				}
+
+				level.Error(m.logger).Log("msg", "Error while verifying token",
+					"err", err.Error())
+				http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (ntk148v) [inline middleware/middleware.go]: IMO should return `http.StatusBadRequest` -> check here https://github.com/dgrijalva/jwt-go/blob/master/parser.go#L23

Comment (vtdat) [reply middleware/middleware.go]: thank you

File: api/cloud.go (hunk 2 above)
Comment (vtdat) [reply api/cloud.go]: ok fixed

File: api/auth.go (hunk 1 above)
Comment (vtdat) [reply api/auth.go]: sure

**Pull request merged**